Only widen short local stores

SingleAccretion opened the pull request:
As the commit message points out, widening byte stores is not profitable most of the time.

This change is meant to minimize the amount of regressions for the upcoming enablement of folding primitives in local morph.

[Diffs](https://github.com/dotnet/runtime/pull/74685/checks?check_run_id=8337332205).

SingleAccretion committed:
Only widen 2 byte wide local stores

// Widening byte -> int:
//
// Zero     : 3 -> 4/2
// Non-Zero : 3 -> 6
//
// Notable: only the "reused zero" case can become better.
//
// Widening short -> int:
//
// Zero     : 5 -> 4/2
// Non-Zero : 5 -> 6 (no 0x66 prefix)
//
// Notable: the zero (common) case is always better, while
// for the non-zero case we only regress by one byte and
// get rid of the prefix.

SingleAccretion committed:
Perhaps this...?

SingleAccretion commented:
@dotnet/jit-contrib

AndyAyersMS commented:
@jakobbotsch can you review?

jakobbotsch commented:
/azp run runtime-coreclr superpmi-diffs

jakobbotsch commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -42,61 +42,29 @@ void Lowering::LowerRotate(GenTree* tree)
 {
-    // Try to widen the ops if they are going into a local var.
-    if ((storeLoc->gtOper == GT_STORE_LCL_VAR) && (storeLoc->gtOp1->gtOper == GT_CNS_INT))
+    // Most small locals (the exception is dependently promoted fields) have 4 byte wide stack slots, so
+    // we can widen the store, if profitable. This optimization is not relevant for register candidates.
```
ARM has the same "optimization" (which does not make much sense on ARM I think). Should it be adjusted? It does not have the `lvDoNotEnregister` heuristic check, but it probably makes sense to remove it entirely.

SingleAccretion replied:
> ARM

I've investigated this question a while back.

1) For ARM64 this widening is useless. I'll `#ifdef` it out and delete the LA64 copy.
2) For ARM, it can be useful because 4-byte-wide stores can be encoded in 2 bytes in more cases than the small ones. I'll make the code look more like this new x86/x64 version.

> It does not have the lvDoNotEnregister heuristic check, but it probably makes sense to remove it entirely.

Yes... Not sure why I put that in; it's true that the "optimization" is only relevant for in-memory locals, but even candidates can end up spilled. I'll remove it.

SingleAccretion committed:
Delete the widening from LA64

On LA64, all instructions have the same size
of 4 bytes, so there is no point to widening.

SingleAccretion committed:
Remove the DNER check

No point.

SingleAccretion committed:
ARM/64 lowering

No point in widening on ARM64.
